src/cargo/ops/cargo_rustc/mod.rs

```diff
@@ -383,6 +355,44 @@ fn rustc(cx: &mut Context, unit: &Unit) -> CargoResult<Work> {
     }
 }
 
+/// Link the compiled target (often of form foo-{metadata_hash}) to the
+/// final target. This must happen during both "Fresh" and "Compile"
+fn link_targets(cx: &mut Context, unit: &Unit) -> CargoResult<Work> {
+    let filenames = try!(cx.target_filenames(unit));
+    Ok(Work::new(move |_| {
+        // If we're a "root crate", e.g. the target of this compilation, then we
+        // hard link our outputs out of the `deps` directory into the directory
+        // above. This means that `cargo build` will produce binaries in
+        // `target/debug` which one probably expects.
+        for (src, link_dst, _linkable) in filenames {
+            // This may have been a `cargo rustc` command which changes the
+            // output, so the source may not actually exist.
+            debug!("Thinking about linking {} to {:?}", src.display(), link_dst);
+            if !src.exists() || link_dst.is_none() {
+                continue
+            }
+            let dst = link_dst.unwrap();
+
+            debug!("linking {} to {}", src.display(), dst.display());
+            if dst.exists() {
+                try!(fs::remove_file(&dst).chain_error(|| {
+                    human(format!("failed to remove: {}", dst.display()))
```

Comment:
That's an interesting hypothesis. I suppose that when running cargo build twice concurrently as in that test, the second build used to hit "fresh" without having to try to access the output. (It could still race on accessing the fingerprint files, but less likely). Now fresh also accesses the output.